### ISSUE TITLE
Args not working in tasks.json using - "_runner":"terminal"

### ISSUE THREAD

**Issue by gdxn96** [Microsoft/vscode#28036]
- VSCode Version: 1.12.2
- OS Version: Windows 10
- Terminal: Bash for Windows
```
{
    "version": "0.1.0",
    "options": {
        "cwd": "${workspaceRoot}"
    },
    "_runner": "terminal",
    "tasks": [
        {
            "taskName": "Install Pip Dependancies",
            "command": ".env/bin/python",
            "isShellCommand": true,
            "args": ["-m", "pip", "install", "-r", "requirements/dev.txt"],
            "showOutput": "always"
        }
    ]
}
```
Running this task does not act as expected, it opens the terminal as it should, then it opens a python interpreter, ignoring all my nice args. Have I done something wrong or is it a bug? I've deleted all my other tasks for legibility.

**Comment by kieferrm** [comment]
I tried that on macOS with 1.12.2 and insiders. Works in both cases without issues.

**Comment by dbaeumer** [comment]
@gdxn96 which Bash system to you use. The Windows Subsystem for Linux ?

**Comment by dbaeumer** [comment]
I can reproduce using Ubuntu bash.exe. The shell args constructed look like this:

![capture](https://cloud.githubusercontent.com/assets/1931590/26818198/00371bf8-4a9a-11e7-99dd-3739fdca7c75.png)

@Tyriar you once told me that under Windows I must pass one string as a shell arg in case of additional arguments. Is there anything else I need to consider. Must the command passed to echo in this case be quoted ?

**Comment by dbaeumer** [comment]
Please note that to easier reproduce I modified the command from .env/bin/python to echo.

**Comment by dbaeumer** [comment]
The code works as expected using cmd.exe or powershell.

**Comment by gdxn96** [comment]
Sorry yes, Windows Subsystem for Linux is whats causing the issue

**Comment by Tyriar** [comment]
@dbaeumer so the issue here seems to be related to us not giving bash.exe what it's expecting. Needs some experimentation.

**Comment by dbaeumer** [comment]
@Tyriar assigning you as well. If you give me some hints I could look into this as well.

**Comment by Tyriar** [comment]
@dbaeumer I would expect either as args or the single string should work. Not sure what's sent now but maybe for bash.exe on Windows you need a conditional that uses the `string[]` approach?

**Comment by giovdk21** [comment]
Hi,
I've the same issue; this is with the insiders build (2017-06-09) on Windows using bash.

```javascript
{
    "version": "2.0.0",
    "tasks": [
        {
            "taskName": "Test 1",
            "type": "shell",
            "command": "uname",
            "isShellCommand": true,
            "echoCommand": true,
            "args": ["-a"],
            "showOutput": "always"    
        }
    ]
}
```

Expected result: `Linux HostName 4.4.0-43-Microsoft #1-Microsoft Wed Dec 31 14:42:53 PST 2014 x86_64 x86_64 x86_64 GNU/Linux`
Actual result: `Linux`

**Comment by dbaeumer** [comment]
@Tyriar I tried both using as single string or a string[]. Both result in the same behavior under Windows of not running the task correctly. What I noticed is that the winpty-agent closed but the electron process started with `--type=terminal` still hangs around afterwards when passing the args separately:

![capture](https://user-images.githubusercontent.com/1931590/27533693-cc1d57ee-5a64-11e7-8ed0-a9273f4f42ec.png)

When passing the args as one string the execution fails with: 
```
The terminal process terminated with exit code: 3221225477
```

Can you please investigate. I ran out of ideas / options.

**Comment by giovdk21** [comment]
For a temporary workaround to this one, please see also: https://github.com/Microsoft/vscode/issues/22663#issuecomment-313818041

**Comment by dbaeumer** [comment]
@Tyriar please see https://github.com/Microsoft/vscode/issues/30417#issue-242153847. It contains a work around by quoting the command line. Is this what I am supposed to do for bash under Windows?

**Comment by 0not** [comment]
@dbaeumer According to the bash man page, the argument to `bash -c` should be quoted. Anything after the string to be executed is passed in as parameters.

```
-c string
If the -c option is present, then commands are read from string. If there are arguments 
after the string, they are assigned to the positional parameters, starting with $0.
```

For example, `"command": "\"echo $0\" Hello"` prints "Hello" as expected. 

It appears that the command does need to be quoted when passed to `bash -c`, but I don't know enough about the task system to say if that's feasible. I just figured I'd pass this information on.

**Comment by giovdk21** [comment]
@dbaeumer & @0not could you please have a look at my previous comment or [here](https://github.com/giovdk21/wsl-bash-tools)?

The workaround I implemented works in a similar way but basically I'm collecting all the shell arguments with a powershell script and then calling `C:\Windows\System32\bash.exe -c "$args_str"` 

This not only fixes it but it also allows for other parameters to be appended at the end. For example one extension (solargraph) is defining a command that it needs to run and then appending different arguments depending on the action it wants to perform.

Another thing to consider is that the path passed from the task / extension is in the windows format, so my workaround also tries to convert that into a linux path.

**Comment by dbaeumer** [comment]
It is correct that the arg passed to -c should be quoted but somehow this is done by the terminal correctly when running bash under Linux and Mac OS but obviously not under Windows. We need to understand what is different in the xterm code and whether quoting it for Windows is the right thing to do.

**Comment by 0not** [comment]
I believe I've found a solution, but I don't know how correct it is. In `src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts` change line 475 from `shellLaunchConfig.args = Platform.isWindows ? shellArgs.join(' ') : shellArgs;` to `shellLaunchConfig.args = Platform.isWindows && basename !== 'bash.exe' ? shellArgs.join(' ') : shellArgs;`. Several lines before this Windows is treated specially with three options (powershell, bash, or cmd), but on this line the only differentiation is made between Windows and other (presumably bash on Linux/Mac). The logic for `isWindows` should really be `not bash`.

I would create a pull request for this, but embarrassingly, I don't know how and I don't have time tonight. I've tested this solution by building vscode on Windows 10, but I don't know of any unforeseen side effects.

**Comment by Tyriar** [comment]
@dbaeumer On Windows you use the different escaping method (`string`), I'm assuming since bash.exe redirects to actual bash on Ubuntu it requires the same format (`string[]`).

You could try see if it's bash.exe `%windir%\(sysnative|system32|syswow64)` then use the `string[]` method?

**Comment by dbaeumer** [comment]
@Tyriar https://github.com/Microsoft/vscode/issues/30417 shows the same problem for bash from git install (so no Linux Subsystem). So should I use string[] for git bash as well. Would be good to understand what the terminal does so I can implement the right fix on my side.

**Comment by Tyriar** [comment]
@dbaeumer all the examples in the issue that introduced the `string` format appear to be cmd.exe (https://github.com/Microsoft/vscode/issues/19078, https://github.com/Tyriar/node-pty/issues/41), maybe it needs to only apply to cmd.exe? I don't have much knowledge into this area other than cmd.exe doesn't want the args in the "proper" argv format. It needs some experimentation to figure out what's going on.

The formatting is done on the `string | string[]` [here](https://github.com/Tyriar/node-pty/blob/50ddb6457d186ad5c77aaa02c867302d8730fb36/src/windowsPtyAgent.ts#L102), after that it's handed off to winpty which is largely a black box to me. That function was added in https://github.com/Tyriar/node-pty/commit/ed3c9da68e4db2981fa1851a7923bd3b522a1d4b, before that arguments didn't work at all on Windows (fixed https://github.com/chjj/pty.js/issues/137).

**Comment by dbaeumer** [comment]
Fixed by fixing #30417
